File: apple/packages/EnteComponents/Sources/EnteComponents/Theme.swift
import CoreText
import SwiftUI

public enum EnteApp {
    case photos
    case auth
    case locker
}

public struct Palette {
    public let isDark: Bool
    public let primary: Color
    public let primaryDark: Color
    public let primaryDarker: Color
    public let primarySurface: Color
    public let text: Color
    public let mutedText: Color
    public let hintText: Color
    public let disabledText: Color
    public let reverseText: Color
    public let background: Color
    public let surface: Color
    public let fill: Color
    public let fillDarker: Color
    public let fillDarkest: Color
    public let border: Color
    public let faintBorder: Color
    public let danger: Color
    public let dangerDark: Color
    public let dangerDarker: Color
    public let caution: Color
    public let information: Color

    static func make(app: EnteApp, scheme: ColorScheme) -> Palette {
        let dark = scheme == .dark
        let primary: Color
        let primaryDark: Color
        let primaryDarker: Color
        let primarySurface: Color

        switch app {
        case .photos:
            primary = Color(hex: 0x08C225)
            primaryDark = Color(hex: 0x069D1E)
            primaryDarker = Color(hex: 0x057C18)
            primarySurface = Color(hex: dark ? 0x292929 : 0xDDEEDF)
        case .auth:
            primary = Color(hex: 0x9610D6)
            primaryDark = Color(hex: 0x7A0CAE)
            primaryDarker = Color(hex: 0x5D0884)
            primarySurface = Color(hex: dark ? 0x271C32 : 0xF4E7FC)
        case .locker:
            primary = Color(hex: 0x1071FF)
            primaryDark = Color(hex: 0x0E5FD9)
            primaryDarker = Color(hex: 0x0B4CAD)
            primarySurface = Color(hex: dark ? 0x292929 : 0xE7EFFA)
        }

        return Palette(
            isDark: dark,
            primary: primary,
            primaryDark: primaryDark,
            primaryDarker: primaryDarker,
            primarySurface: primarySurface,
            text: dark ? .white : .black,
            mutedText: Color(hex: dark ? 0x999999 : 0x666666),
            hintText: Color(hex: 0x969696),
            disabledText: Color(hex: dark ? 0x414141 : 0xD6D6D6),
            reverseText: dark ? .black : .white,
            background: Color(hex: dark ? 0x161616 : 0xF4F4F4),
            surface: Color(hex: dark ? 0x212121 : 0xFFFFFF),
            fill: Color(hex: dark ? 0x0A0A0A : 0xEAEAEA),
            fillDarker: Color(hex: dark ? 0x141414 : 0xDEDEDE),
            fillDarkest: Color(hex: dark ? 0x292929 : 0xD2D2D2),
            border: Color(hex: dark ? 0x3E3E3E : 0xE0E0E0),
            faintBorder: Color(hex: dark ? 0x2A2A2A : 0xEBEBEB),
            danger: Color(hex: 0xF63A3A),
            dangerDark: Color(hex: 0xDD3434),
            dangerDarker: Color(hex: 0xC52E2E),
            caution: Color(hex: 0xF08A1E),
            information: Color(hex: 0x1071FF)
        )
    }
}

public enum EnteSpacing {
    public static let xs: CGFloat = 4
    public static let sm: CGFloat = 8
    public static let md: CGFloat = 12
    public static let lg: CGFloat = 16
    public static let xl: CGFloat = 20
    public static let xxl: CGFloat = 24
}

public enum EnteRadius {
    public static let small: CGFloat = 8
    public static let medium: CGFloat = 12
    public static let large: CGFloat = 16
    public static let button: CGFloat = 20
    public static let sheet: CGFloat = 24
}

public enum EnteIconSize {
    public static let micro: CGFloat = 8
    public static let tiny: CGFloat = 12
    public static let small: CGFloat = 18
    public static let medium: CGFloat = 24
    public static let large: CGFloat = 36
}

public enum EnteMotion {
    public static let quick = 0.12
    public static let standard = 0.18
    public static let slow = 0.26
}

public enum EnteTypography {
    public static let display1 = EnteFont.outfit(size: 32, relativeTo: .largeTitle)
    public static let display2 = EnteFont.outfit(size: 24, relativeTo: .title)
    public static let heading1 = EnteFont.inter(size: 20, weight: .bold, relativeTo: .title2)
    public static let heading2 = EnteFont.inter(size: 18, weight: .semibold, relativeTo: .title3)
    public static let large = EnteFont.inter(size: 16, weight: .semibold, relativeTo: .body)
    public static let body = EnteFont.inter(size: 14, weight: .medium, relativeTo: .body)
    public static let bodyBold = EnteFont.inter(size: 14, weight: .semibold, relativeTo: .body)
    public static let mini = EnteFont.inter(size: 12, weight: .medium, relativeTo: .caption)
    public static let tiny = EnteFont.inter(size: 10, weight: .medium, relativeTo: .caption2)
    public static let avatarExtraSmall = EnteFont.inter(size: 8, weight: .medium, relativeTo: .caption2)
    public static let avatarSmall = EnteFont.inter(size: 10, weight: .medium, relativeTo: .caption2)
}

public struct EnteTheme<Content: View>: View {
    @Environment(\.colorScheme) private var colorScheme
    private let app: EnteApp
    private let content: Content

    public init(app: EnteApp = .photos, @ViewBuilder content: () -> Content) {
        self.app = app
        self.content = content()
    }

    public var body: some View {
        let palette = Palette.make(app: app, scheme: colorScheme)
        content
            .environment(\.entePalette, palette)
            .tint(palette.primary)
    }
}

public extension EnvironmentValues {
    var entePalette: Palette {
        get { self[PaletteKey.self] }
        set { self[PaletteKey.self] = newValue }
    }
}

private struct PaletteKey: EnvironmentKey {
    static let defaultValue = Palette.make(app: .photos, scheme: .light)
}

private enum EnteFont {
    static func inter(size: CGFloat, weight: Font.Weight, relativeTo: Font.TextStyle) -> Font {
        let name = switch weight {
        case .bold: "Inter-Bold"
        case .semibold: "Inter-SemiBold"
        case .medium: "Inter-Medium"
        default: "Inter-Regular"
        }
        return custom(name, size: size, relativeTo: relativeTo)
    }

    static func outfit(size: CGFloat, relativeTo: Font.TextStyle) -> Font {
        custom("Outfit-SemiBold", size: size, relativeTo: relativeTo)
    }

    private static func custom(_ name: String, size: CGFloat, relativeTo: Font.TextStyle) -> Font {
        _ = registration
        return .custom(name, size: size, relativeTo: relativeTo)
    }

    private static let registration: Void = {
        ["Inter-Regular", "Inter-Medium", "Inter-SemiBold", "Inter-Bold", "Outfit-SemiBold"].forEach { name in
            guard let url = Bundle.module.url(forResource: name, withExtension: "ttf") else { return }
            CTFontManagerRegisterFontsForURL(url as CFURL, .process, nil)
        }
    }()
}

extension Color {
    init(hex: UInt) {
        self.init(
            red: Double((hex >> 16) & 0xFF) / 255,
            green: Double((hex >> 8) & 0xFF) / 255,
            blue: Double(hex & 0xFF) / 255
        )
    }
}

File: android/packages/components/src/main/kotlin/io/ente/components/Theme.kt
package io.ente.components

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.ColorScheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

public enum class EnteApp { Photos, Auth, Locker }

@Immutable
public class EntePalette internal constructor(
    public val isDark: Boolean,
    public val primary: Color,
    public val primaryDark: Color,
    public val primaryDarker: Color,
    public val primarySurface: Color,
    public val text: Color,
    public val mutedText: Color,
    public val hintText: Color,
    public val disabledText: Color,
    public val reverseText: Color,
    public val background: Color,
    public val surface: Color,
    public val fill: Color,
    public val fillDarker: Color,
    public val fillDarkest: Color,
    public val border: Color,
    public val faintBorder: Color,
    public val danger: Color,
    public val dangerDark: Color,
    public val dangerDarker: Color,
    public val caution: Color,
    public val information: Color,
)

public val LocalEntePalette: ProvidableCompositionLocal<EntePalette> = staticCompositionLocalOf {
    palette(EnteApp.Photos, dark = false)
}

public object EnteSpacing {
    public val xs: Dp = 4.dp
    public val sm: Dp = 8.dp
    public val md: Dp = 12.dp
    public val lg: Dp = 16.dp
    public val xl: Dp = 20.dp
    public val xxl: Dp = 24.dp
}

public object EnteRadius {
    public val small: Dp = 8.dp
    public val medium: Dp = 12.dp
    public val large: Dp = 16.dp
    public val button: Dp = 20.dp
    public val sheet: Dp = 24.dp
}

public object EnteIconSize {
    public val micro: Dp = 8.dp
    public val tiny: Dp = 12.dp
    public val small: Dp = 18.dp
    public val medium: Dp = 24.dp
    public val large: Dp = 36.dp
}

public object EnteMotion {
    public const val quick: Int = 120
    public const val standard: Int = 180
    public const val slow: Int = 260
}

public object EnteTypography {
    private val inter: FontFamily = FontFamily(
        Font(R.font.ente_inter_regular, FontWeight.Normal),
        Font(R.font.ente_inter_medium, FontWeight.Medium),
        Font(R.font.ente_inter_semibold, FontWeight.SemiBold),
        Font(R.font.ente_inter_bold, FontWeight.Bold),
    )
    private val outfit: FontFamily = FontFamily(
        Font(R.font.ente_outfit_semibold, FontWeight.SemiBold),
    )

    public val display1: TextStyle = style(outfit, 32.sp, FontWeight.SemiBold, 40.sp)
    public val display2: TextStyle = style(outfit, 24.sp, FontWeight.SemiBold, 32.sp)
    public val heading1: TextStyle = style(inter, 20.sp, FontWeight.Bold, 28.sp)
    public val heading2: TextStyle = style(inter, 18.sp, FontWeight.SemiBold, 24.sp)
    public val large: TextStyle = style(inter, 16.sp, FontWeight.SemiBold, 20.sp)
    public val body: TextStyle = style(inter, 14.sp, FontWeight.Medium, 20.sp)
    public val bodyBold: TextStyle = style(inter, 14.sp, FontWeight.SemiBold, 20.sp)
    public val mini: TextStyle = style(inter, 12.sp, FontWeight.Medium, 16.sp)
    public val tiny: TextStyle = style(inter, 10.sp, FontWeight.Medium, 12.sp)
    public val avatarExtraSmall: TextStyle = style(inter, 8.sp, FontWeight.Medium, 15.sp)
    public val avatarSmall: TextStyle = style(inter, 10.sp, FontWeight.Medium, 15.sp)

    internal val material: Typography = Typography(
        headlineLarge = display1,
        headlineMedium = display2,
        titleLarge = heading1,
        titleMedium = heading2,
        bodyLarge = body,
        bodyMedium = body,
        labelLarge = bodyBold,
        labelMedium = mini,
        labelSmall = tiny,
    )

    private fun style(
        fontFamily: FontFamily,
        fontSize: TextUnit,
        fontWeight: FontWeight,
        lineHeight: TextUnit,
    ): TextStyle = TextStyle(
        fontFamily = fontFamily,
        fontSize = fontSize,
        fontWeight = fontWeight,
        lineHeight = lineHeight,
    )
}

@Composable
public fun EnteTheme(
    app: EnteApp = EnteApp.Photos,
    darkTheme: Boolean = isSystemInDarkTheme(),
    content: @Composable () -> Unit,
) {
    val palette = palette(app, darkTheme)
    androidx.compose.runtime.CompositionLocalProvider(LocalEntePalette provides palette) {
        MaterialTheme(
            colorScheme = palette.colorScheme(),
            typography = EnteTypography.material,
            content = content,
        )
    }
}

private fun EntePalette.colorScheme(): ColorScheme = if (isDark) {
    darkColorScheme(
        primary = primary,
        onPrimary = reverseText,
        secondary = primarySurface,
        onSecondary = primary,
        background = background,
        onBackground = text,
        surface = surface,
        onSurface = text,
        surfaceVariant = fill,
        onSurfaceVariant = mutedText,
        outline = border,
        error = danger,
    )
} else {
    lightColorScheme(
        primary = primary,
        onPrimary = reverseText,
        secondary = primarySurface,
        onSecondary = primary,
        background = background,
        onBackground = text,
        surface = surface,
        onSurface = text,
        surfaceVariant = fill,
        onSurfaceVariant = mutedText,
        outline = border,
        error = danger,
    )
}

private fun palette(app: EnteApp, dark: Boolean): EntePalette {
    val primary = when (app) {
        EnteApp.Photos -> Color(0xFF08C225)
        EnteApp.Auth -> Color(0xFF9610D6)
        EnteApp.Locker -> Color(0xFF1071FF)
    }
    val primaryDark = when (app) {
        EnteApp.Photos -> Color(0xFF069D1E)
        EnteApp.Auth -> Color(0xFF7A0CAE)
        EnteApp.Locker -> Color(0xFF0E5FD9)
    }
    val primaryDarker = when (app) {
        EnteApp.Photos -> Color(0xFF057C18)
        EnteApp.Auth -> Color(0xFF5D0884)
        EnteApp.Locker -> Color(0xFF0B4CAD)
    }
    val primarySurface = when (app) {
        EnteApp.Photos -> if (dark) Color(0xFF292929) else Color(0xFFDDEEDF)
        EnteApp.Auth -> if (dark) Color(0xFF271C32) else Color(0xFFF4E7FC)
        EnteApp.Locker -> if (dark) Color(0xFF292929) else Color(0xFFE7EFFA)
    }

    return EntePalette(
        isDark = dark,
        primary = primary,
        primaryDark = primaryDark,
        primaryDarker = primaryDarker,
        primarySurface = primarySurface,
        text = if (dark) Color.White else Color.Black,
        mutedText = Color(if (dark) 0xFF999999 else 0xFF666666),
        hintText = Color(0xFF969696),
        disabledText = Color(if (dark) 0xFF414141 else 0xFFD6D6D6),
        reverseText = if (dark) Color.Black else Color.White,
        background = Color(if (dark) 0xFF161616 else 0xFFF4F4F4),
        surface = if (dark) Color(0xFF212121) else Color.White,
        fill = Color(if (dark) 0xFF0A0A0A else 0xFFEAEAEA),
        fillDarker = Color(if (dark) 0xFF141414 else 0xFFDEDEDE),
        fillDarkest = Color(if (dark) 0xFF292929 else 0xFFD2D2D2),
        border = Color(if (dark) 0xFF3E3E3E else 0xFFE0E0E0),
        faintBorder = Color(if (dark) 0xFF2A2A2A else 0xFFEBEBEB),
        danger = Color(0xFFF63A3A),
        dangerDark = Color(0xFFDD3434),
        dangerDarker = Color(0xFFC52E2E),
        caution = Color(0xFFF08A1E),
        information = Color(0xFF1071FF),
    )
}
